packages/core/src/lib/lit-core.ts

```diff
@@ -381,30 +397,166 @@ export class LitCore {
+    if (this._connectingPromise) {
+      return this._connectingPromise;
+    }
+
+    this._connectingPromise = this._connect();
+
+    await this._connectingPromise.finally(() => {
+      this._connectingPromise = null;
+    });
+  }
+
+  private async _connect() {
+    // Ensure an ill-timed epoch change event doesn't trigger concurrent config changes while we're already doing that
+    this._stopListeningForNewEpoch();
+
+    // Ensure we don't fire an existing network sync poll handler while we're in the midst of connecting anyway
+    this._stopNetworkPolling();
+
     await this.setNewConfig();
+
+    // -- handshake with each node.  Note that if we've previously initialized successfully, but this call fails,
+    // core will remain useable but with the existing set of `connectedNodes` and `serverKeys`.
+    const { connectedNodes, serverKeys, coreNodeConfig } =
+      await this._runHandshakeWithBootstrapUrls();
+    Object.assign(this, { ...coreNodeConfig, connectedNodes, serverKeys });
+
+    this.scheduleNetworkSync();
     await this.listenForNewEpoch();
-    await this._runHandshakeWithBootstrapUrls();
-  };
 
-  /**
+    // FIXME: don't create global singleton; multiple instances of `core` should not all write to global
+    // @ts-ignore
+
+    globalThis.litNodeClient = this;
+    this.ready = true;
+
+    log(`🔥 lit is ready. "litNodeClient" variable is ready to use globally.`);
+    log('current network config', {
+      networkPubkey: this.networkPubKey,
+      networkPubKeySet: this.networkPubKeySet,
+      hdRootPubkeys: this.hdRootPubkeys,
+      subnetPubkey: this.subnetPubKey,
+      latestBlockhash: this.latestBlockhash,
+    });
+
+    // browser only
+    if (isBrowser()) {
+      document.dispatchEvent(new Event('lit-ready'));
+    }
+  }
+
+  private async handshakeAndVerifyNodeAttestation({
+    url,
+    requestId,
+  }: {
+    url: string;
+    requestId: string;
+  }): Promise<JsonHandshakeResponse> {
+    const challenge = this.getRandomHexString(64);
+
+    const handshakeResult = await this.handshakeWithNode(
+      { url, challenge },
+      requestId
+    );
+
+    const keys: JsonHandshakeResponse = {
+      serverPubKey: handshakeResult.serverPublicKey,
+      subnetPubKey: handshakeResult.subnetPublicKey,
+      networkPubKey: handshakeResult.networkPublicKey,
+      networkPubKeySet: handshakeResult.networkPublicKeySet,
+      hdRootPubkeys: handshakeResult.hdRootPubkeys,
+      latestBlockhash: handshakeResult.latestBlockhash,
+    };
+
+    // Nodes that have just bootstrapped will not have negotiated their keys, yet
+    // They will return ERR for those values until they reach consensus
+
+    // Note that if node attestation checks are disabled or checkSevSnpAttestation() succeeds, we will still track the
+    // node, even though its keys may be "ERR".
+    // Should we really track servers with ERR as keys?
+    if (
+      keys.serverPubKey === 'ERR' ||
+      keys.subnetPubKey === 'ERR' ||
```

Comment:
We should also check if they are `undefined` or `null`

Reply:
Looking at the Rust node code, it looks like these _must always be strings_ (see https://github.com/LIT-Protocol/lit-assets/blob/a7b2cbe01fd89384b1579a31a0bc2acc3a7d19ba/rust/lit-node/src/models/mod.rs#L248)

... should we also check for empty strings? I considered making that change actually, but I noticed that `serverPubKey` is **_always_** an empty string, so if we do, it will make the logging very spammy as we will _always_ log errors when running handshakes against Cayenne. Happy to change this validation however you think makes the most sense; I didn't change the behaviour in this PR, just moved it into a method dedicated to performing the handshake and validating the node attestation. Let me know!

Reply:
yeah i think this is fine for now - it's the same behavior as before